File: python/nuclei_segmentation_training/resave_data_for_training.py
from glob import glob
import numpy as np
from tifffile import imread
import h5py
from os.path import split
import os
from scipy.ndimage import zoom
from scipy.ndimage import binary_dilation
from scipy.ndimage import binary_erosion
from scipy.ndimage import binary_closing
from skimage.morphology import area_opening
from skimage.morphology import remove_small_objects
import matplotlib.pyplot as plt
import sys

sys.path.insert(0, "../utils")
from norm_percentile_nocrop import norm_percentile_nocrop






src_path = '../../data_zenodo/part1/nucleus_segmentation'
dst_hdf5_file = '../../data_zenodo/part1_resaved/nucleus_segmentation.hdf5'



img_filenames = glob(src_path + '/**/data_*.tif', recursive=True)
masks_filenames = [ split(x)[0] + '/' + split(x)[1].replace('data_','mask_') for x in img_filenames]
img_numbers = [ split(x)[1].replace('data_','').replace('.tif','') for x in img_filenames]


resized_img_size = [505, 681, 48] #image is resized to this size

normalization_percentile = 0.0001  #image is normalized into this percentile range

mask_erosion=[14, 14, 5] # amount of mask erosion (elipsoid)
minimal_nuclei_size = 6000


X,Y,Z = np.meshgrid(np.linspace(-1,1,mask_erosion[0]),np.linspace(-1,1,mask_erosion[1]),np.linspace(-1,1,mask_erosion[2]))
sphere = np.sqrt(X**2 + Y**2 + Z**2) < 1



if not os.path.exists(split(dst_hdf5_file)[0]):
    os.makedirs(split(dst_hdf5_file)[0])
    
    
with h5py.File(dst_hdf5_file, "w") as f:
    
    for file_num,(img_filename,mask_filename,img_number) in enumerate(zip(img_filenames,masks_filenames,img_numbers)):
        
        print(str(file_num) + ' / ' + str(len(img_filenames)))
        
        if  file_num == 77:
            continue
        
        
        img = imread(img_filename)  
        img = np.moveaxis(img,0,2)

        mask = imread(mask_filename)
        
        
        factor =  np.array(resized_img_size) / np.array(mask.shape)
        mask = zoom(mask,factor,order=0)
        
        
        tmp_size = resized_img_size.copy()
        tmp_size.append(img.shape[3])
        img_resized = np.zeros(tmp_size,dtype=np.float32)
        for channel in range(img.shape[3]):
            
            
            data_one_channel = img[...,channel]
            
            data_one_channel = zoom(data_one_channel,factor,order=1)
            data_one_channel = norm_percentile_nocrop(data_one_channel,normalization_percentile);
        
            img_resized[...,channel] = data_one_channel
            
        img = img_resized
            
            
        
        mask_erroded = np.zeros(mask.shape,dtype=bool);
        for nuclei_value in range(1,5):
            
            print(nuclei_value)
            
            mask_current = mask == nuclei_value;
    
            mask_current = binary_erosion(binary_closing(mask_current,sphere),sphere)
    
            mask_erroded[mask_current] = True


        mask_erroded = remove_small_objects(mask_erroded,minimal_nuclei_size)     
        
        
        
        img_filename = os.path.normpath(img_filename)
        train_valid_test_str = img_filename.split(os.sep)[-2]
        
        
        f.create_dataset(train_valid_test_str + '/mask/' + img_number, data=mask_erroded, chunks=(128,128,48), compression="gzip", compression_opts=2)
        
        f.create_dataset(train_valid_test_str + '/img/' + img_number, data=img_resized,chunks=(128,128,48,3), compression="gzip", compression_opts=2)
        
        
        
        
        
        
        
    
    
    
    
    

File: matlab_nets/nuclei_segmentation_training/resave_data_for_training.m
clc;clear all;close all force;
addpath('../utils')


%% setup
src_path = '../../data_zenodo/part1/nucleus_segmentation';
dst_paht = '../../data_zenodo/part1_resaved/nucleus_segmentation';


% load filenames of data
filenames_imgs = subdir([src_path '/data_*.tif']);
filenames_imgs = {filenames_imgs(:).name};

filenames_masks = {};
for file_num = 1:length(filenames_imgs)
    filename_img = filenames_imgs{file_num};
    [filepath,name,ext] = fileparts(filename_img);
    name(1:4) = 'mask';
    filename_mask = [filepath, '/',name, ext];
    filenames_masks = [filenames_masks filename_mask];
end


resized_img_size = [505  681   48]; %image is resized to this size

normalization_percentile = 0.0001;  %image is normalized into this percentile range

mask_erosion=[14 14 5]; % amount of mask erosion (elipsoid)
minimal_nuclei_size = 6000;

%% resaving 
for img_num = 1:length(filenames_imgs)

    filename_img = filenames_imgs{img_num};
    filename_mask = filenames_masks{img_num};

    data_all_channels = imread_4D_tif(filename_img);


    data = {};
    %% data_preprocessing
    for channel_num = 1:size(data_all_channels,4)

        data_one_channel = data_all_channels(:,:,:,channel_num);

        data_one_channel = imresize3(data_one_channel,resized_img_size);

        data_one_channel = norm_percentile_nocrop(data_one_channel,normalization_percentile);

        data_one_channel = single(data_one_channel);

        data = [data,data_one_channel];
    end

    data = cat(4,data{:});
    
    output_filename = filename_img;
    output_filename = replace(output_filename,'.tif','.mat');
    output_filename = replace(norm_path(output_filename),norm_path(src_path),norm_path(dst_paht));

    mkdir(fileparts(output_filename))

    save(output_filename,'data','-v7.3')


    mask = imread(filename_mask);

    mask = imresize3(mask,resized_img_size,'nearest');

    [X,Y,Z] = meshgrid(linspace(-1,1,mask_erosion(1)),linspace(-1,1,mask_erosion(2)),linspace(-1,1,mask_erosion(3)));
    sphere=sqrt(X.^2+Y.^2+Z.^2)<1;
    

    % preprocess masks - nuclei contains values 1-4 
    mask_erroded = false(size(mask));
    for nuclei_value = 1:4
        mask_current = mask == nuclei_value;

        mask_current = imerode(imclose(mask_current,sphere),sphere); %close to make smooth border
        mask_current = bwareaopen(mask_current,minimal_nuclei_size);

        mask_erroded(mask_current) = true;
    end

    data = mask_erroded;

    output_filename = filename_mask;
    output_filename = replace(output_filename,'.tif','.mat');
    output_filename = replace(norm_path(output_filename),norm_path(src_path),norm_path(dst_paht));

    mkdir(fileparts(output_filename))

    save(output_filename,'data','-v7.3')

end